### DISCUSSION TITLE
Is `Host: ${ip}` a valid request?

### DISCUSSION THREAD

**Discussion by ioquatix** [puma/puma#2977] [General]
We haven incoming requests with a borked host header, i.e.

```
Host: ${ip}
```

(literally that string)

My question is, who should be responsible for rejecting this?

The server?

Clearly the host header is "invalid" if you are to try and construct a URL from it.

**Comment by MSP-Greg** [comment]
> Clearly the host header is "invalid" if you are to try and construct a URL from it.

Glad you checked that! 😁

Probably a messy regex to fully check it. Maybe check for invalid characters (not location sensitive)?  Not.sure.

**Reply by ioquatix** [reply]
`URI` has this:

```
URI::Parser.new.regexp[:HOST]
```

If we did make this a recommendation (or requirement), would we respond with 400 bad request if it didn't match?

**Reply by ioquatix** [reply]
Actually that pattern does not fully match the host header. We might need to put a recommendation in Rack, e.g. `Rack::HOST_PATTERN` or something.

**Comment by MSP-Greg** [comment]
I found a reasonable place to check it.  Checking for a fully valid host name with URI slowed small responses (1kB body, 1kB headers) almost 5%.  Checking that the value contained only valid characters (I used `/[^a-zA-Z0-9:.\[\]\-]/`) seems to add no more than 1% to the RPS numbers.

Really not sure who/what/where should check it.  I didn't track down the actual spec, but several references seemed to imply that an 'invalid' host should generate a 400.  I guess a distinction is whether it's an invalid host or **the wrong host**.

**Reply by ioquatix** [reply]
I'd be totally fine with the approach that achieves 90% of what we want and leaves actual validation up to the app.

That being said, if there was a cheap way to write `valid_host?(host_header)` I'd be all for that, because I think this would mitigate a lot of weird injection/redirect/rebinding attacks.

**Comment by dentarg** [comment]
@ioquatix are you not using Rails? https://github.com/rails/rails/issues/29893#issuecomment-467591014 (or do you need to handle a dynamic list of hosts?)

**Reply by ioquatix** [reply]
I think this was a result of pen testing a Rails app, and the error was because we had some route handling/middleware that was blowing up trying to parse the host header using `URI`.

**Comment by MSP-Greg** [comment]
I can't remember much from ~ 18 months ago, so I need to see if I have a branch testing this, etc.

The regex above is okay, I think it might be better to check for a valid IP address or a valid name, as the regex seems to combine them. 

Maybe the best thing to do is add host validation as an optional feature?  Does anyone have a IETF reference handy?  I've got all the specs, but haven't searched thru them for a while...

**Reply by dentarg** [reply]
I don't think we need or should build any validation into Puma. See https://github.com/rack/rack/issues/1970

**Reply by MSP-Greg** [reply]
Servers may have the benefit of validating the header before a request body is processed?

**Reply by dentarg** [reply]
Yeah, sure. I guess that's a good thing when it comes to stopping garbage requests. https://github.com/rack/rack/issues/1970 has some RFC links.

**Comment by dentarg** [comment]
@nateberkopec added this as an issue: https://github.com/puma/puma/issues/3436
